### PR TITLE
Camera Controller: expose complete API to python

### LINKED ISSUE #70: Improve camera control
The current camera control uses vanilla `Qt3DExtras::QOrbitCameraController` and should be improved.  The camera control is a critical component in the viewer.  It at least needs the following functions and more:

* Resetting.
* WSAD.
* Change of origin.

----

Amended on 27th July 2024

- [x] "Up" and "w" should move toward the top of the window.  "Down" and "s" should move toward the bottom of the window.  "Q" and "e" should change to move along the z axis (perpendicular to the window).
  - I am not sure about the binding to page up/down.  We can use discussions.
  - Implemented with PR #408.
- [x] Camera reset/original position to be configurable and exposed through Python API.  More information: https://github.com/solvcon/modmesh/pull/408#discussion_r1718535001
  - Done with PR #418

### PR DESCRIPTION
**Pull request by Tucchhaa**
In this PR:
* New props added to camera controller to control it's reset behaviour: defaultPosition, defaultUpVector, defaultViewCenter, defaultLookSpeed, defaultLinearSpeed
* All these props are exposed to python
* Camera controller python API was rewritten to use `def_property` using macros. Tests also were updated
* Minor issue was fixed: When reset camera, aspect ratio was incorrect

### CODE REVIEW

File: cpp/modmesh/view/R3DWidget.cpp [hunk 1]
```diff
@@ -99,23 +100,6 @@ void R3DWidget::resizeEvent(QResizeEvent * event)
...
-    camera->setUpVector(QVector3D(0.f, 1.f, 0.f));
-
-    // Set up the camera control.
-    auto * control = m_scene->controller();
-    control->setCamera(camera);
-    control->setLinearSpeed(50.0f);
-    control->setLookSpeed(180.0f);
-}
-
 } /* end namespace modmesh */
```
Comment (Tucchhaa) [inline cpp/modmesh/view/R3DWidget.cpp:103]: resetCamera moved to CameraController

File: cpp/modmesh/view/R3DWidget.hpp [hunk 2]
```diff
@@ -73,7 +71,7 @@ class RScene
 
 private:
 
-    Qt3DExtras::QAbstractCameraController * m_controller = nullptr;
+    CameraController * m_controller;
```
Comment (Tucchhaa) [inline cpp/modmesh/view/R3DWidget.hpp:74]: small refactoring to `controller()` getters. Logic didn't change, but because our `CameraController` class is used more often than Qt's less `dynamic_cast`'s will happen

File: cpp/modmesh/view/R3DWidget.cpp [hunk 3]
```diff
@@ -99,23 +100,6 @@ void R3DWidget::resizeEvent(QResizeEvent * event)
...
-    camera->setViewCenter(QVector3D(0.0f, 0.0f, 0.0f));
-    camera->setUpVector(QVector3D(0.f, 1.f, 0.f));
-
-    // Set up the camera control.
-    auto * control = m_scene->controller();
-    control->setCamera(camera);
-    control->setLinearSpeed(50.0f);
-    control->setLookSpeed(180.0f);
-}
-
```
Comment (Tucchhaa) [inline cpp/modmesh/view/R3DWidget.cpp:118]: `resetCamera` moved to `CameraController`

File: cpp/modmesh/view/R3DWidget.hpp [hunk 4]
```diff
@@ -46,24 +46,22 @@
...
-    Qt3DExtras::QAbstractCameraController const * controller() const { return m_controller; }
-    Qt3DExtras::QAbstractCameraController * controller() { return m_controller; }
+    CameraController * controller() const { return m_controller; }
+    Qt3DExtras::QAbstractCameraController * qtController() const { return m_controller->asQtCameraController(); }
 
-    void setCameraController(Qt3DExtras::QAbstractCameraController * controller)
+    void setCameraController(CameraController * controller)
     {
-        m_controller->deleteLater();
+        qtController()->deleteLater();
```
Comment (Tucchhaa) [inline cpp/modmesh/view/R3DWidget.hpp]: small refactoring to `controller()` getters. Logic didn't change, but because our `CameraController` class is used more often than Qt's less `dynamic_cast`'s will happen

File: cpp/modmesh/view/RCameraController.cpp [hunk 5]
```diff
@@ -212,6 +212,18 @@ void RCameraInputListener::initKeyboardListeners() const
     m_ty_axis->addInput(m_keyboard_ty_neg_input);
 }
 
+void CameraController::reset()
+{
+    camera()->lens()->setPerspectiveProjection(45.0f, camera()->lens()->aspectRatio(), 0.1f, 1000.0f);
```
Comment (Tucchhaa) [inline cpp/modmesh/view/RCameraController.cpp]: Now don't reset `aspectRatio`. Because if resize a window and reset, then mesh will look strange. Qt automatically already updates `aspectRatio` when window is resized

File: cpp/modmesh/view/RCameraController.hpp [hunk 6]
```diff
@@ -135,30 +135,59 @@ class CameraController
...
+    void setDefaultViewCenter(QVector3D value) { m_default_view_center = value; }
+
+    QVector3D defaultUpVector() const { return m_default_up_vector; }
+    void setDefaultUpVector(QVector3D value) { m_default_up_vector = value; }
+
+    float defaultLinearSpeed() const { return m_default_linear_speed; }
+    void setDefaultLinearSpeed(float value) { m_default_linear_speed = value; }
+
+    float defaultLookSpeed() const { return m_default_look_speed; }
+    void setDefaultLookSpeed(float value) { m_default_look_speed = value; }
```
Comment (Tucchhaa) [inline cpp/modmesh/view/RCameraController.hpp:183]: new props in `CameraController` to control reset behaviour

File: cpp/modmesh/view/wrap_view.cpp [hunk 7]
```diff
@@ -460,8 +432,7 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapRCameraController
                     bool left_mouse_button,
                     bool right_mouse_button,
                     bool alt_key,
-                    bool shift_key,
-                    float dt)
```
Comment (Tucchhaa) [inline cpp/modmesh/view/wrap_view.cpp:464]: remove `dt` parameter, because no need of it in python.

File: cpp/modmesh/view/wrap_view.cpp [hunk 8]
```diff
@@ -196,39 +196,7 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapR3DWidget
...
-
-        (*this)
-            // clang-format off
-            DECL_QVECTOR3D_PROPERTY(position, position, setPosition)
-            DECL_QVECTOR3D_PROPERTY(up_vector, upVector, setUpVector)
-            DECL_QVECTOR3D_PROPERTY(view_center, viewCenter, setViewCenter)
-            // clang-format on
-            ;
-
-#undef DECL_QVECTOR3D_PROPERTY
```
Comment (Tucchhaa) [inline cpp/modmesh/view/wrap_view.cpp:231]: this is now available through `widget.camera`

File: cpp/modmesh/view/wrap_view.cpp [hunk 9]
```diff
@@ -487,44 +459,64 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapRCameraController
...
+
+        (*this)
+            // clang-format off
+                DECL_FLOAT_PROPERTY(linear_speed, linearSpeed, setLinearSpeed)
+                DECL_FLOAT_PROPERTY(look_speed, lookSpeed, setLookSpeed)
+                DECL_FLOAT_PROPERTY(default_linear_speed, defaultLinearSpeed, setDefaultLinearSpeed)
+                DECL_FLOAT_PROPERTY(default_look_speed, defaultLookSpeed, setDefaultLookSpeed)
+            // clang-format on
+            ;
+#undef DECL_FLOAT_PROPERTY
```
Comment (Tucchhaa) [inline cpp/modmesh/view/wrap_view.cpp]: Now use macro instead of defining each getter and setter

File: tests/test_view.py [hunk 10]
```diff
@@ -111,190 +103,231 @@ def normalize(self, vec):
...
+
+        c.reset()
+
+        self.assertEqual(c.position, (1, 2, 3))
+        self.assertEqual(c.view_center, (4, 5, 6))
+        self.assertEqual(c.up_vector, (7, 8, 9))
+        self.assertEqual(c.linear_speed, 123.0)
+        self.assertEqual(c.look_speed, 456.0)
+
+
```
Comment (Tucchhaa) [inline tests/test_view.py]: test common things of both camera types. For now it's default properties

File: tests/test_view.py [hunk 11]
```diff
@@ -111,190 +103,231 @@ def normalize(self, vec):
...
-        self.widget.resetCamera()
+        self.camera.reset()
 
     def test_reset(self):
-        dt = 0.01
-        initial_pos = self.pos()
-        initial_view_vector = self.view_vector()
-        initial_view_center = self.view_center()
-        initial_up_vector = self.up_vector()
+        c = self.camera
```
Comment (Tucchhaa) [inline tests/test_view.py:206]: refactored all camera tests:
* removed usage of `dt` in `camera.move` function
* use properties instead of getters and setters

**Comment by Tucchhaa** [comment]
@yungyuc ready for review

File: cpp/modmesh/view/RCameraController.cpp (hunk 5 above)
Comment (tigercosmos) [reply cpp/modmesh/view/RCameraController.cpp]: Better to have a comment to explain what is 45.0, 0.1, and 1000.0. Or, you can have `constexpr float angle = 45.0f`, which can better explain the code.

File: cpp/modmesh/view/RCameraController.hpp [hunk 12]
```diff
@@ -135,30 +135,59 @@ class CameraController
...
     Qt3DExtras::QAbstractCameraController * asQtCameraController()
     {
         return dynamic_cast<Qt3DExtras::QAbstractCameraController *>(this);
     }
+
+protected:
+    RCameraInputListener * m_listener = nullptr;
+
+private:
+    QVector3D m_default_position = QVector3D(0.0f, 0.0f, 10.0f);
```
Comment (tigercosmos) [inline cpp/modmesh/view/RCameraController.hpp:189]: Why (0, 0, 10)?

File: cpp/modmesh/view/wrap_view.cpp [hunk 13]
```diff
@@ -487,44 +459,64 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapRCameraController
...
+                    DECL_QVECTOR3D_PROPERTY(up_vector, upVector, setUpVector)
+                    DECL_QVECTOR3D_PROPERTY(view_center, viewCenter, setViewCenter)
+                    DECL_QVECTOR3D_PROPERTY(default_position, defaultPosition, setDefaultPosition)
+                    DECL_QVECTOR3D_PROPERTY(default_view_center, defaultViewCenter, setDefaultViewCenter)
+                    DECL_QVECTOR3D_PROPERTY(default_up_vector, defaultUpVector, setDefaultUpVector)
+            // clang-format on
+            ;
+#undef DECL_QVECTOR3D_PROPERTY
+
+#define DECL_FLOAT_PROPERTY(NAME, GETTER, SETTER) \
```
Comment (tigercosmos) [inline cpp/modmesh/view/wrap_view.cpp]: The macro naming doesn't follow the style rules.
https://github.com/solvcon/modmesh/blob/master/STYLE.rst#c-pre-processor-macro

File: tests/test_view.py [hunk 14]
```diff
@@ -111,190 +103,231 @@ def normalize(self, vec):
...
-        self.assertEqual(view_center[1], new_view_center[1])
-        self.assertEqual(view_center[2], new_view_center[2])
+        self.assertEqual(c.view_center[0], new_view_center[0])
+        self.assertEqual(c.view_center[1], new_view_center[1])
+        self.assertEqual(c.view_center[2], new_view_center[2])
 
     def test_rotation(self):
-        dt = 0.01
-        angle = dt * self.look_speed()
+        c = self.camera
```
Comment (tigercosmos) [inline tests/test_view.py:226]: Personally, I don't like a shorthand like `c`, but maybe yyc has a different thought.

File: cpp/modmesh/view/R3DWidget.cpp (hunk 3 above)
Comment (yungyuc) [reply cpp/modmesh/view/R3DWidget.cpp:118]: Moving it to `CameraController` makes it more testable.  I like it.

File: cpp/modmesh/view/R3DWidget.hpp [hunk 15]
```diff
@@ -46,24 +46,22 @@
 namespace modmesh
 {
 
-class RScene
-    : public Qt3DCore::QEntity
+class RScene : public Qt3DCore::QEntity
```
Comment (yungyuc) [inline cpp/modmesh/view/R3DWidget.hpp]: Did clang-format force you to make it in one line: `class RScene : public Qt3DCore::QEntity`?

If not, please change it back to use the two lines and avoid this unnecessary change of code.

File: cpp/modmesh/view/R3DWidget.hpp [hunk 16]
```diff
@@ -46,24 +46,22 @@
...
 {
-
 public:
 
-    explicit RScene(Qt3DCore::QNode * parent = nullptr)
-        : Qt3DCore::QEntity(parent)
-        , m_controller(new ROrbitCameraController(this))
+    explicit RScene(QNode * parent = nullptr)
+        : QEntity(parent)
```
Comment (yungyuc) [inline cpp/modmesh/view/R3DWidget.hpp:55]: It's good to separate the `CameraController` from the scene.

File: cpp/modmesh/view/R3DWidget.hpp [hunk 17]
```diff
@@ -46,24 +46,22 @@
...
     {
+        m_controller = new ROrbitCameraController(this);
     }
 
-    Qt3DExtras::QAbstractCameraController const * controller() const { return m_controller; }
-    Qt3DExtras::QAbstractCameraController * controller() { return m_controller; }
+    CameraController * controller() const { return m_controller; }
+    Qt3DExtras::QAbstractCameraController * qtController() const { return m_controller->asQtCameraController(); }
```
Comment (yungyuc) [inline cpp/modmesh/view/R3DWidget.hpp]: Why do we need an helper `qtController()` for casting?  Is there a reason that the consumers cannot do it themselves?

```cpp
Qt3DExtras::QAbstractCameraController * qtc = scene->controller()->asQtCameraController();
```

If you want to avoid `Qt3DExtras::QAbstractCameraController const *`, an overload of `CameraController * controller()` should work.

File: cpp/modmesh/view/R3DWidget.hpp (hunk 4 above)
Comment (yungyuc) [reply cpp/modmesh/view/R3DWidget.hpp]: Is there a reason why we do not make (our) `CameraController` a sub class of `QAbstractCameraController`?

File: cpp/modmesh/view/RCameraController.cpp (hunk 5 above)
Comment (yungyuc) [reply cpp/modmesh/view/RCameraController.cpp]: It is a good idea to add local `constexpr` for constants.  It will be better document than comments.

Angles need special treatment.  If it's in degree, the (local) name should contain it like `front_deg`.  If it's in radian, the name should contain it accordingly.

It's impossible to exclusively use either degree or radian for angles.  Both are needed by different libraries.  Conversion is a must.  Local variable names should make it clear which unit is used.

---

Names for member data and function parameters of angles are trickier.

File: cpp/modmesh/view/wrap_view.cpp (hunk 9 above)
Comment (yungyuc) [reply cpp/modmesh/view/wrap_view.cpp]: Long live C pre-processor.  It is very handy in C++.

File: tests/test_view.py [hunk 18]
```diff
@@ -30,6 +30,7 @@
 import os
 
 import modmesh
+
```
Comment (yungyuc) [inline tests/test_view.py]: Please remove the unnecessary change of blank line.

File: tests/test_view.py [hunk 19]
```diff
@@ -111,190 +103,231 @@ def normalize(self, vec):
...
+        c.move(yaw=1, left_mouse_button=True)
 
         rotation_matrix = self.angle_axis(angle, (0, 1, 0))
         rotated_vector = np.array(initial_view_vector) @ rotation_matrix
-        view_vector = self.view_vector()
 
-        self.assertAlmostEqual(rotated_vector[0], view_vector[0], delta=1e-5)
-        self.assertAlmostEqual(rotated_vector[1], view_vector[1], delta=1e-5)
-        self.assertAlmostEqual(rotated_vector[2], view_vector[2], delta=1e-5)
+        self.assertAlmostEqual(rotated_vector[0], c.view_vector[0], places=5)
```
Comment (yungyuc) [inline tests/test_view.py]: `delta` is clearer than `places`.  Why change to the latter?

File: tests/test_view.py [hunk 20]
```diff
@@ -111,190 +103,231 @@ def normalize(self, vec):
         return vec / np.linalg.norm(vec)
 
 
+@unittest.skipIf(GITHUB_ACTIONS, "GUI is not available in GitHub Actions")
+class ViewCommonCameraTC(ViewCameraTB, unittest.TestCase):
+    camera_type = "fps"  # no difference when use orbit camera
```
Comment (yungyuc) [inline tests/test_view.py]: I believe this comment suggests they behave the same in theory.  If that is true, both fps and orbit cameras should be tested.  Testing is required to make sure the theory is always the reality.

You may use the TB/TC pattern to test both.

**Review by yungyuc — CHANGES_REQUESTED** [review]
> In this PR:
> 
> * New props added to camera controller to control it's reset behaviour: defaultPosition, defaultUpVector, defaultViewCenter, defaultLookSpeed, defaultLinearSpeed
> * All these props are exposed to python
> * Camera controller python API was rewritten to use `def_property` using macros. Tests also were updated
> * Minor issue was fixed: When reset camera, aspect ratio was incorrect

The new Python wrapper and tests are great.

It is unfortunate we cannot execute the tests with github actions.  It's an improvement we should make in the future.

---

- [x] Clarify why [`Qt3DExtras::QAbstractCameraController * qtController() const`](https://github.com/solvcon/modmesh/pull/418/files#diff-f940faded53f4d9cc5445a00082fe9d330c8afddcafef0ba4a42b51d6a77f0e5R60) is needed.
- [x] Is there a reason why we do not make (our) `CameraController` a sub class of `QAbstractCameraController`?  (See discussion at [`setCameraController()`](https://github.com/solvcon/modmesh/pull/418/files#diff-f940faded53f4d9cc5445a00082fe9d330c8afddcafef0ba4a42b51d6a77f0e5R64)).
- [x] Fix the naming of [`#define DECL_FLOAT_PROPERTY(NAME, GETTER, SETTER)`](https://github.com/solvcon/modmesh/pull/418/files#diff-88bff1c42c8197e364f9cbfb23de0593b395fc90d08bd52b8c0f0a7a05ab03beR499): Follow the macro naming style https://github.com/solvcon/modmesh/blob/master/STYLE.rst#c-pre-processor-macro
- [x] Clarify why change `delta` to `places` for [`assertAlmostEqual()`](https://github.com/solvcon/modmesh/pull/418/files#diff-f50af349a1f116824ed7af1180dc8a287127e38fca8ae87d9b556053aafbba44R220)?
- [x] Test both [fps and orbit](https://github.com/solvcon/modmesh/pull/418/files#diff-f50af349a1f116824ed7af1180dc8a287127e38fca8ae87d9b556053aafbba44R108) cameras.

File: tests/test_view.py (hunk 19 above)
Comment (Tucchhaa) [reply tests/test_view.py]: Because flake8's limitation to 80 characters makes me do weird things 😵. `places=5` is two characters less than`delta=1e-5`

File: cpp/modmesh/view/R3DWidget.hpp (hunk 4 above)
Comment (Tucchhaa) [reply cpp/modmesh/view/R3DWidget.hpp]: I like your question. Before I intended to not repeat functionality of Qt in our code. But now since I already duplicated almost all properties to expose it to python, we can remove mixin and make it as you suggest.

File: cpp/modmesh/view/R3DWidget.hpp (hunk 17 above)
Comment (Tucchhaa) [reply cpp/modmesh/view/R3DWidget.hpp]: I'll make `CameraController` inherit `QAbstractCameraController`, so `qtController` will be removed

File: cpp/modmesh/view/RCameraController.hpp (hunk 12 above)
Comment (Tucchhaa) [reply cpp/modmesh/view/RCameraController.hpp:189]: Because meshes are placed at `(0, 0, 0)`, and we don't want the camera to be inside the mesh

**Comment by Tucchhaa** [comment]
In last commit:
* Used constexpr for 'magic values' in `reset` function ([code](https://github.com/solvcon/modmesh/pull/418/files#diff-be9511fa234431d29c196ed57eaf68dbf67a702bc5400c605cdb9d5450a48d33R217-R219))
* Made `CameraController` to inherit from `QAbstractCameraController` ([code](https://github.com/solvcon/modmesh/pull/418/files#diff-d22d8f512997b8a57a09a8179b7446c525a92a5797d54e5e8dfa0922555814bbR129)) and removed all now unnecessary functions, like `qtCameraControler`. And getters and setters from FPS and Orbit subclasses moved to `CameraController`
* Fixed formatting issues
* Test for both fps and orbit camera types in `ViewCommonCameraTC` ([code](https://github.com/solvcon/modmesh/pull/418/files#diff-f50af349a1f116824ed7af1180dc8a287127e38fca8ae87d9b556053aafbba44R108))

@yungyuc @tigercosmos ready for your review 🙂

File: cpp/modmesh/view/RCameraController.hpp (hunk 12 above)
Comment (tigercosmos) [reply cpp/modmesh/view/RCameraController.hpp:189]: I can guess that, but my point is, better to have some comments for it.

Comment (yungyuc) [reply cpp/modmesh/view/RCameraController.hpp:189]: >> Because meshes are placed at `(0, 0, 0)`, and we don't want the camera to be inside the mesh

To put the camera outside the mesh, it is not good to hard-code the default value of the member `m_default_position`, because the mesh may be placed anywhere.

We do not have a reasonably large number of mesh examples for the default location to show disadvantage.

> I can guess that, but my point is, better to have some comments for it.

I concur, but it's OK to leave it here and deal with it in the future.  As we make the code more useful, some meshes will cover the position sooner or later.  We will immediately know and fix it.

File: cpp/modmesh/view/RCameraController.hpp [hunk 21]
```diff
@@ -128,77 +126,95 @@ class RCameraInputListener : public Qt3DCore::QEntity
     Qt3DInput::QButtonAxisInput * m_keyboard_tz_neg_input;
 }; /* end class RCameraInputListener */
 
-class CameraController
+class CameraController : public Qt3DExtras::QAbstractCameraController
```
Comment (yungyuc) [inline cpp/modmesh/view/RCameraController.hpp]: This needs to be named as `RCameraController` because it is a Qt class customized (derived) in modmesh.  See https://github.com/solvcon/modmesh/blob/master/STYLE.rst#qt for the modmesh Qt style.

File: cpp/modmesh/view/RCameraController.hpp [hunk 22]
```diff
@@ -128,77 +126,95 @@ class RCameraInputListener : public Qt3DCore::QEntity
...
+
 public:
-    virtual ~CameraController() = default;
+    explicit CameraController(Qt3DCore::QNode * parent = nullptr)
+        : QAbstractCameraController(parent)
+    {
+    }
 
-    virtual void updateCameraPosition(const CameraInputState & state, float dt) = 0;
+    // Do nothing in QAbstractCameraController's moveCamera
```
Comment (yungyuc) [inline cpp/modmesh/view/RCameraController.hpp]: Thanks for the informative comment.  It will work even better to put the comment in the function body:

```cpp
{
    // Do nothing in QAbstractCameraController's moveCamera
}
```

File: tests/test_view.py (hunk 19 above)
Comment (yungyuc) [reply tests/test_view.py]: That's a clear and valid reason.  But it is also unfortunate 🤣.

Could you please rename `rotated_vector` as `rot_vec` or even `rvec`, and change back to use `delta=1.e-5`?  Using `delta` gives us more correct behaviors.  We will copy and paste the code for future testing.  Behaviorally correct code gives us longer mileage.

We cannot change the argument name nor the float literals, but modmesh developers have the constitutive right to rename local variables.  (We have not written how to treat local variable names in the [style guide](https://github.com/solvcon/modmesh/blob/4166c1aa6ffca26f5cd5865606d495d366ddc9e2/STYLE.rst), but we should.)

File: tests/test_view.py [hunk 23]
```diff
@@ -111,190 +104,235 @@ def normalize(self, vec):
...
+            self.assertEqual(c.position, (1, 2, 3))
+            self.assertEqual(c.view_center, (4, 5, 6))
+            self.assertEqual(c.up_vector, (7, 8, 9))
+            self.assertEqual(c.linear_speed, 123.0)
+            self.assertEqual(c.look_speed, 456.0)
+
+
 @unittest.skipIf(GITHUB_ACTIONS, "GUI is not available in GitHub Actions")
 class ViewFPSCameraTC(ViewCameraTB, unittest.TestCase):
     camera_type = "fps"
```
Comment (yungyuc) [inline tests/test_view.py:200]: The testing code in `ViewFPSCameraTC` and `ViewOrbitCameraTC` looks very similar if not the same.  Is it possible to merge to a base class (`*TB`)?

**Review by yungyuc — CHANGES_REQUESTED** [review]
Thanks for clarifying the many things swiftly.  You updated pretty quickly!  Not sure what time zone you are in 🙂

Additional requests after that:
- [x] [`RCameraController.hpp:129`](https://github.com/solvcon/modmesh/pull/418/files#diff-d22d8f512997b8a57a09a8179b7446c525a92a5797d54e5e8dfa0922555814bbR129): Rename `CameraController` to `RCameraController`.
- [x] [`test_view.py:225`](https://github.com/solvcon/modmesh/pull/418/files#diff-f50af349a1f116824ed7af1180dc8a287127e38fca8ae87d9b556053aafbba44R225): Rename `rotated_vector` as `rot_vec` or something like `rvec`, and change back to use `delta=1.e-5`.
- [x] [`test_view.py:158`](https://github.com/solvcon/modmesh/pull/418/files#diff-f50af349a1f116824ed7af1180dc8a287127e38fca8ae87d9b556053aafbba44R158): Is it possible to merge the testing code in `ViewFPSCameraTC` and `ViewOrbitCameraTC` to a base class (`*TB`)?  (If it is too cumbersome to do it now or you don't like a `*TB`, we can leave the duplicated code here for now.)

File: cpp/modmesh/view/R3DWidget.hpp (hunk 4 above)
Comment (yungyuc) [reply cpp/modmesh/view/R3DWidget.hpp]: Either way works for me.  Qt prefers inheritance, but HPC code usually uses templates and/or mixins.  There is probably not a definite answer for everything.  But for the Qt code here, inheritance might not be wrong.

File: tests/test_view.py (hunk 23 above)
Comment (Tucchhaa) [reply tests/test_view.py:200]: `test_rest`, `test_rotation` are actually different and `test_translation` indeed is the same in both fps and orbit. But it's not easy to move common tests to the base class, because unittest can't understand that structure.

I will move `test_translation` to `ViewCommonCameraTC`

**Comment by Tucchhaa** [comment]
> You updated pretty quickly! Not sure what time zone you are in 🙂

It's only 2 hours earlier for me. Btw, I will arrive to Taiwan very soon)

File: tests/test_view.py (hunk 14 above)
Comment (Tucchhaa) [reply tests/test_view.py:226]: I did that due to flake8's rule to limit lines to 80 characters. If I write `self.camera` or `camera` everywhere I will get a lot of errors and I will have to shorten more other variables :/

Comment (yungyuc) [reply tests/test_view.py:226]: A local `c` in test code is good.  It can't be right to skim through testing code for a complex structure for 3D camera with only a glance.  Writing `c = self.camera` forces a maintainer to pause for a second to keep the information in their brain's short-term memory block, and improves their understanding of the code.

**Comment by yungyuc** [comment]
> > You updated pretty quickly! Not sure what time zone you are in 🙂
> 
> It's only 2 hours earlier for me. Btw, I will arrive to Taiwan very soon)

Around the end of the summer vacation one can be anywhere.  I hope you had a good time this summer.  Having said that, we will be more than happy to share some good ideas of traveling!

File: tests/test_view.py (hunk 23 above)
Comment (yungyuc) [reply tests/test_view.py:200]: It's good to move identical testing code to a single place.  Thanks.

---

A side note.  For testing, it is not rare to duplicate code intentionally.  Compared to "feature" code, we more often change the testing code for experimenting without checking in the change (we want the testing code to be slightly more stable than the feature code).

In this PR, I do not see a reason to duplicate testing code.  This is just FYI.

**Comment by Tucchhaa** [comment]
In last PR:
* Renamed `CameraController` to `RCameraController`
* moved duplicated test function `test_translate` to `ViewCommonCameraTC`
* Replaced `places=5` to `delta=1e-5`

**Comment by Tucchhaa** [comment]
@yungyuc 
I have a question. Does it matter how we name commits? I try to give them meaningful names, but sometimes it's hard to do that. Like last to commits are just 'apply review' and 'apply review 2'. Is it ok for modmesh?

**Comment by yungyuc** [comment]
I'll review the code after the CI completes.

> In last PR:

I think you meant to say commit (or update to the PR).  But I am nitpicking.  If you like to write accurately, it's a reminder.  Whatever term you use does not bother me.

> @yungyuc I have a question. Does it matter how we name commits? I try to give them meaningful names, but sometimes it's hard to do that. Like last to commits are just 'apply review' and 'apply review 2'. Is it ok for modmesh?

It depends.  For a non-trivial enhancement like this PR, some mouthy commits are fine.  But in general I do not want to be picky on commit logs.  I guess I am "just reasonably picky".  We'll see.

File: cpp/modmesh/view/wrap_view.cpp [hunk 24]
```diff
@@ -487,44 +459,64 @@ class MODMESH_PYTHON_WRAPPER_VISIBILITY WrapRCameraController
...
+            // clang-format off
+                MM_DECL_FLOAT_PROPERTY(linear_speed, linearSpeed, setLinearSpeed)
+                MM_DECL_FLOAT_PROPERTY(look_speed, lookSpeed, setLookSpeed)
+                MM_DECL_FLOAT_PROPERTY(default_linear_speed, defaultLinearSpeed, setDefaultLinearSpeed)
+                MM_DECL_FLOAT_PROPERTY(default_look_speed, defaultLookSpeed, setDefaultLookSpeed)
+            // clang-format on
+            ;
+#undef MM_DECL_FLOAT_PROPERTY
     }
 };
```
Comment (yungyuc) [inline cpp/modmesh/view/wrap_view.cpp:521]: Class ending mark is not here.  It should be added in the future.

**Review by yungyuc — APPROVED** [review]
LGTM to me.  Missed class ending marks (e.g., [`wrap_view.cpp:521`](https://github.com/solvcon/modmesh/pull/418/files#diff-88bff1c42c8197e364f9cbfb23de0593b395fc90d08bd52b8c0f0a7a05ab03beR521)) may be addressed in the future.
